Feature/stash support

orclev opened the pull request:
Proposed implementation for #38.

cburgdorf commented:
Mmmh...I wonder if we can come up with something more generic. Something like a concept of a "link pattern" and then people could make that work with stash, github, beanstalk or whatever. 

What if we allow a `link-pattern` configuration that takes a string pattern which would default to `https://github.com/{user}/{repo}/commit/{commit-hash}`. If we go that route we could deprecate our current `repository` config in favor for the more fine grained `user` and `repo` configs. And then `{user}`, `{repo}`, `{commit-hash}` would just become usable placeholders within the `link-pattern` configuration.

How does that sound to you?

//cc @kbknapp

kbknapp commented:
I'd say that's a good longer-term goal. I'm pushing a PR to this PR to remove some boiler plate provided by `clap` and unstable features. Once those are merged, I'd say this is good step in that direction and should be fully compatible (from a user perspective) with those future goals.

kbknapp commented:
I forgot to add, unstable features aren't that big a deal for a binary...just in this particular case I think we don't lose anything by keeping everything on stable.

orclev commented:
I merged your PR, do I need to issue a new PR with those changes included?

On the topic of using a pattern I like that idea in principle, but I do have some concern. In particular with Stash some of those pieces don't exactly line up. E.G. there isn't necessarily a user. One of our repo links takes the form `http://{server}:{port}/projects/{projectName}/repos/{repoName}/commits/{commitHash}`

kbknapp commented:
Nope, it's automatic :)

orclev commented:
I can rebase on master once I'm home (I'm at work currently).

kbknapp commented on src/clogconfig.rs:
```diff
@@ -4,6 +4,7 @@ use std::path::Path;
 use std::fmt::Display;
 use std::env;
 use std::collections::HashMap;
+use std::ascii::AsciiExt;
```
Just saw this, it's not necessary since the `arg_enum!{}` macro also expands to include this.

orclev commented:
Alright, rebased it onto the latest master branch. Once the CI finishes I think it should be good to merge (I also cleaned up that one issue you spotted).

kbknapp commented:
Looks good to me :+1:

kbknapp commented:
I've been putting some thought into this, and I think for now the way @orclev has done it is the best way. One change I did decide would be better though; I'd change `repoflavor` or `link-style` because this is slightly more clear about the intent.

If we were to go down the full `link-pattern` route now, I don't think it'd actually be too big of a gain over what's already in this PR. Because the base URL is specified in `--repository` and the rest is taken care of by `--link-style` (or whatever name). What we currently (after this PR) would have is almost more generic because things like the one @orclev stated would work just fine and we don't even need to worry about parsing a URL pattern. 

I'm not super familiar with all the different styles of online `git` repos, but so long as they all *end* the same (i.e. the `{repo}` is the last thing before the commit/issue links), this PR works great.

One last note, I believe gitlab uses the same github style links, if you'd like to add that (even though functionally it's it's a nop, I just wouldn't want a user saying "Well I guess they only support github and stash, so my gitlab servers our out").

orclev commented:
I'll make those changes sometime in the next couple days.

kbknapp commented:
:+1: 
@cburgdorf I'm good with this, if you are since it's your call

cburgdorf commented:
Sorry, I'm on a cycling trip with tent and sleeping bag. Things like electric power and connectivity are limited for me right now so I can't engage much. 

I agree with you that we can follow this route for now. Feel free to merge.

kbknapp commented:
:+1: Merged.

Sounds like fun, I'm jealous!

cburgdorf commented:
![20150528_182525](https://cloud.githubusercontent.com/assets/521109/7886953/e154a500-062d-11e5-988a-c5598d6d68fa.jpg)

Yes, you should be ;)
